src/main/java/io/prometheus/wls/rest/UrlBuilder.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2020, Oracle Corporation and/or its affiliates.
+ *
+ * Licensed under the Universal Permissive License v 1.0 as shown at http://oss.oracle.com/licenses/upl.
+ */
+
+package io.prometheus.wls.rest;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import javax.servlet.http.HttpServletRequest;
+
+import io.prometheus.wls.rest.domain.Protocol;
+
+public class UrlBuilder {
+
+  private static final List<Integer> successes = new ArrayList<>();
+
+  private final String host;
+  private final Protocol protocol;
+  private final Queue<Integer> ports = new PriorityQueue<>(new PreferSuccesses());
+  private int lastCandidate;
+
+  public UrlBuilder(HttpServletRequest request, Integer restPort) {
+    host = request.getLocalName();
+    protocol = Protocol.getProtocol(request);
+    Optional.ofNullable(restPort).ifPresent(ports::add);
+    addPortIfNotPresent(request.getLocalPort());
+  }
+
+  private void addPortIfNotPresent(int port) {
+    if (!ports.contains(port)) ports.add(port);
+  }
+
+  public static void clearHistory() {
+    successes.clear();
+  }
+
+  public String createUrl(String urlPattern) {
+    return protocol.format(urlPattern, host, selectPort());
+  }
+
+  private int selectPort() {
+    if (ports.isEmpty()) throw new WebClientException("No connection port is defined");
+
+    return lastCandidate = ports.peek();
+  }
+
+  // Informs the builder that the last URL it supplied worked
+  public void reportSuccess() {
+    successes.add(lastCandidate);
```

Comment:
would this keep adding new elements to the `successes` array if all url requests are successful?

Reply:
Good point, thanks. That should be a set.